From 864bf72edbf66a94e9eb9abc6d7a9d3e16428c6c Mon Sep 17 00:00:00 2001
From: Lu Yahan <yahan@iscas.ac.cn>
Date: Mon, 13 Feb 2023 10:36:25 +0800
Subject: [PATCH] [riscv32]Fix regress-1412940 DCHECK failed.

Using UseScratchRegisterScope to get a temp reg in Spill.



Change-Id: I6c333dc6cb6bcfe0f7c4f716831532d8a4a18d73
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/4243114
Reviewed-by: ji qiu <qiuji@iscas.ac.cn>
Auto-Submit: Yahan Lu <yahan@iscas.ac.cn>
Commit-Queue: Yahan Lu <yahan@iscas.ac.cn>
Cr-Commit-Position: refs/heads/main@{#85782}
---
 .../baseline/riscv/liftoff-assembler-riscv32.h | 18 ++++++++----------
 1 file changed, 8 insertions(+), 10 deletions(-)

diff --git a/src/wasm/baseline/riscv/liftoff-assembler-riscv32.h b/src/wasm/baseline/riscv/liftoff-assembler-riscv32.h
index 63def4f7149..0cf8825391c 100644
--- a/src/wasm/baseline/riscv/liftoff-assembler-riscv32.h
+++ b/src/wasm/baseline/riscv/liftoff-assembler-riscv32.h
@@ -1001,25 +1001,23 @@ void LiftoffAssembler::Spill(int offset, LiftoffRegister reg, ValueKind kind) {
 void LiftoffAssembler::Spill(int offset, WasmValue value) {
   RecordUsedSpillOffset(offset);
   MemOperand dst = liftoff::GetStackSlot(offset);
+  UseScratchRegisterScope assembler_temps(this);
+  Register tmp = assembler_temps.Acquire();
   switch (value.type().kind()) {
     case kI32:
     case kRef:
     case kRefNull: {
-      LiftoffRegister tmp = GetUnusedRegister(kGpReg, {});
-      MacroAssembler::li(tmp.gp(), Operand(value.to_i32()));
-      Sw(tmp.gp(), dst);
+      MacroAssembler::li(tmp, Operand(value.to_i32()));
+      Sw(tmp, dst);
       break;
     }
     case kI64: {
-      LiftoffRegister tmp = GetUnusedRegister(kGpRegPair, {});
-
       int32_t low_word = value.to_i64();
       int32_t high_word = value.to_i64() >> 32;
-      MacroAssembler::li(tmp.low_gp(), Operand(low_word));
-      MacroAssembler::li(tmp.high_gp(), Operand(high_word));
-
-      Sw(tmp.low_gp(), liftoff::GetHalfStackSlot(offset, kLowWord));
-      Sw(tmp.high_gp(), liftoff::GetHalfStackSlot(offset, kHighWord));
+      MacroAssembler::li(tmp, Operand(low_word));
+      Sw(tmp, liftoff::GetHalfStackSlot(offset, kLowWord));
+      MacroAssembler::li(tmp, Operand(high_word));
+      Sw(tmp, liftoff::GetHalfStackSlot(offset, kHighWord));
       break;
       break;
     }
-- 
2.35.1

